### PR TITLE
WIP File Tree API

### LINKED ISSUE #10: Provide directory-helper-methods
For my libstorj.NET-wrapper I want to provide methods to Access Folders and subfolders within buckets. There is an ongoing discussion about this [as a SIP here](https://github.com/Storj/sips/issues/19).

I Can do something within my wrapper, but then you would Need to do it on your side, too, and we have code-duplication with probably different implementation.

I suggest extending your API with the following methods:

GetContent(Bucket b)
=> should return direct subfolders and files within the bucket as e.g. a "Content-object".
=> a Content-object might have two lists: files and folders

Get Content(Folder f)
=> shoudl return direct subfolders and files within a Folder as e.g. a "Content-object" as described above.

Once the SIP is implemented, the Internat structure Can be mapped accordingly. For the current implementation this might also make a buffer necessary, so that the Content of a bucket does not get reloaded on every API-Access.

### PR DESCRIPTION
**Pull request by kaloyan-raev**
Fixes #10

### CODE REVIEW

**Comment by TopperDEL** [comment]
It is working so far!

But I would still propose to return a generic "Content"-object - that Can be either a file or a Folder. Havin a file-object with the property "isDirectory" feels wrong. But that is just my opinion and not a big deal.

What I also would suggest is to put the methods "getChildren" into another helper-object to seperate that logic from the base-storj-functions. So if we introduce a Kind of "StorjExtended"-class or "StorjDirectoryHelper" it would make it more clear what happens here. On constructing such an object, you could take the original storj-object as Parameter so you Can call the base-functions.
Can you follow me? I'm Feeling like I cannot express myself currently. :)

But for the Moment, this helps me alot developing my next-project, the StorjFS filesystem. :) Thank you!

**Comment by kaloyan-raev** [comment]
One of the options I considered was indeed a higher level API that calls the current low level API. It's good that you share the same thoughts, so I will explore this direction in my next iteration today.

**Comment by TopperDEL** [comment]
High Level API and Lov Level API - These were the words I was missing yesterday. :)

Yes, I would suggest that, too.

**Comment by kaloyan-raev** [comment]
I've pushed the reworked API. Look at the `FileTreeExample` for usage.

**Comment by TopperDEL** [comment]
Looks good! Will have a deeper look at it - but might not happen before easter. But the coding and structures Looks good. Thank you for your effort and help!

File: src/main/java/io/storj/libstorj/fs/Dir.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import io.storj.libstorj.Bucket;
+import io.storj.libstorj.Entry;
+import io.storj.libstorj.File;
+
+@SuppressWarnings("serial")
+public class Dir extends Entry {
+
+    String bucketId;
+
+    Dir(Entry entry) {
```
Comment (TopperDEL) [inline src/main/java/io/storj/libstorj/fs/Dir.java]: Both constructors Need to be public, as I cannot use them on my wrapper.

**Review by TopperDEL — CHANGES_REQUESTED** [review]
Both constructors of class Dir Need to be public - otherwise I cannot wrap them correctly.

**Comment by TopperDEL** [comment]
We Need to buffer the entries, though. Maybe if you just load the bucket-Contents once until a Change-Operation is performed?

**Comment by TopperDEL** [comment]
Seems I found an error - but I'm not completely sure (it's late and I'm tired :) ).

I have a bucket called "Filezilla2" with the following Content (exemplary):
"Filezilla2\Richtfest\Norah Jones\Come Away With Me\Painter Song.mp3"

If I move from Filezilla2 to Richtfest and should get the conent "Norah Jones" - I get "Norah Jones\Come Away With Me\" instead. So your logic seems to jump over one Directory-depth. Maybe you can reproduce that.

**Comment by RichardLitt** [comment]
👋 Hey! Thanks for this contribution. Apologies for the delay in responding!

We've decided to rearchitect Storj, so that we can scale better. You can read more about this decision [here](https://storj.io/blog/2018/01/getting-from-petabytes-to-exabytes-the-road-ahead/). This means that we are entirely focused on v3 at the moment, in the [storj/storj](https://github.com/storj/storj) repository. Our white paper for v3 is coming **very, very** soon - follow along on [the blog](https://storj.io/blog) and in our [Rocketchat](https://community.storj.io/home).

As this repository is part of the v2 network, we're no longer maintaining this repository. I am going to close this for now. If you have any questions, I encourage you to jump on Rocketchat and ask them there. Thanks!

**Pull request closed without merging**